### PR TITLE
Further completion API refactoring

### PR DESCRIPTION
**Pull request by DustinCampbell**
Putting this out there to get reviewers looking at it. It's not yet complete, but some of the tangles in this code are better. Reviewers: @CyrusNajmabadi, @jasonmalinowski, @dpoeschl, @rchande, @Pilchie.

### CODE REVIEW

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs [hunk 1]
```diff
@@ -4,8 +4,9 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using Microsoft.CodeAnalysis.Completion;
-using Microsoft.CodeAnalysis.Completion.Providers;
+using Microsoft.CodeAnalysis.Completion.Triggers;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs]: Should this be a separate namespace?  Seems so tightly tied to Completion that it should be in the same namespace.  Not sure.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs [hunk 2]
```diff
@@ -162,12 +177,16 @@ private bool StartNewModelComputation(ICompletionService completionService, Comp
 
             this.sessionOpt = new Session(this, computation, GetCompletionRules(), Presenter.CreateSession(TextView, SubjectBuffer, null));
 
-            var completionProviders = triggerInfo.TriggerReason == CompletionTriggerReason.Snippets
+            var completionProviders = trigger is DisplaySnippetsCompletionTrigger
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs]: should we just use a well known name for this case?  this seems pretty hacky having to know exactly the class that was instantiated.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Backspace.cs [hunk 3]
```diff
@@ -98,7 +99,7 @@ private void ExecuteBackspaceOrDelete(ITextView textView, Action nextHandler, bo
                     this.StopModelComputation();
                     return;
                 }
-                else if (model != null && model.TriggerInfo.TriggerReason != CompletionTriggerReason.BackspaceOrDeleteCommand)
+                else if (model != null && !(model.Trigger is BackspaceOrDeleteCharCompletionTrigger))
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Backspace.cs]: same here.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs [hunk 4]
```diff
@@ -135,8 +135,8 @@ private void Commit(CompletionItem item, TextChange textChange, Model model, cha
                 }
             }
 
-            // Let the completion rules know that this item was committed.
-            GetCompletionRules().CompletionItemCommitted(item);
+            // Ensure that the item is added to the completion service's most recently used list.
+            GetCompletionService().MostRecentlyUsedList.AddItem(item);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs]: is this something internal?  I'm a little wary of making mutable state public.

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/AbstractCSharpCompletionProviderTests.cs [hunk 5]
```diff
@@ -133,7 +134,7 @@ protected void VerifySendEnterThroughToEnter(string initialMarkup, string textTy
                 var document = workspace.CurrentSolution.GetDocument(documentId);
                 var position = hostDocument.CursorPosition.Value;
 
-                var completionList = GetCompletionList(document, position, CompletionTriggerInfo.CreateInvokeCompletionTriggerInfo());
+                var completionList = GetCompletionList(document, position, new DisplayListCompletionTrigger());
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/AbstractCSharpCompletionProviderTests.cs]: as this has no args, why not have a static readonly instance of it?  You do that later on for other triggers it seems lke.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Backspace.cs [hunk 6]
```diff
@@ -63,10 +63,11 @@ private void ExecuteBackspaceOrDelete(ITextView textView, Action nextHandler, bo
                     this.TextView.Caret.PositionChanged += OnCaretPositionChanged;
                 }
 
-                var triggerInfo = CompletionTriggerInfo.CreateBackspaceTriggerInfo(deletedChar);
+                var trigger = new BackspaceOrDeleteCharCompletionTrigger(deletedChar, isBackspace: !isDelete, customTags: _completionTriggerTags);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Backspace.cs]: seems odd.  why not have different triggers for each?

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TabKey.cs [hunk 7]
```diff
@@ -82,7 +83,7 @@ private bool TryInvokeSnippetCompletion(TabKeyCommandArgs args, ICompletionServi
                         {
                             var textChange = new TextChange(TextSpan.FromBounds(caretPoint - 1, caretPoint), string.Empty);
                             workspace.ApplyTextChanges(documentId, textChange, CancellationToken.None);
-                            this.StartNewModelComputation(completionService, CompletionTriggerInfo.CreateSnippetTriggerInfo(), filterItems: false);
+                            this.StartNewModelComputation(completionService, DisplaySnippetsCompletionTrigger.Instance, filterItems: false);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TabKey.cs]: Here you have a static instance.  Earlier you were instantiating.  I'd like this to be consistent.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/DescriptionModifyingCompletionItem.cs [hunk 8]
```diff
@@ -53,6 +54,24 @@ public override CompletionListProvider CompletionProvider
...
         }
 
+        /// <summary>
+        /// Returns a string that should be added to the description of the given completion item
+        /// if a snippet exists with a shortcut that matches the completion item's insertion text.
+        /// </summary>
+        private Task<string> GetSnippetExpansionNoteForCompletionItemAsync(CompletionItem completionItem)
+        {
+            var completionRules = _completionService.GetCompletionRules();
+            var insertionText = completionRules.GetTextChange(completionItem, '\t').NewText;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/DescriptionModifyingCompletionItem.cs]: I'm not sure what this means.  What's the \t for?

File: src/Features/Core/Completion/AbstractCompletionService.cs [hunk 9]
```diff
@@ -8,94 +8,57 @@
...
-                    _committedItems.RemoveAt(0);
-                }
-
-                _committedItems.Add(item.DisplayText);
-            }
-        }
+        public string LanguageName { get; }
+        public MostRecentlyUsedList MostRecentlyUsedList { get; } = new MostRecentlyUsedList();
 
         internal int GetMRUIndex(CompletionItem item)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Completion/AbstractCompletionService.cs]: are these necessary if we're now exposing the MRUList directly?

File: src/Features/Core/Completion/AbstractCompletionService.cs [hunk 10]
```diff
@@ -8,94 +8,57 @@
...
-            return candidate;
+            this.LanguageName = languageName;
         }
 
         public abstract IEnumerable<CompletionListProvider> GetDefaultCompletionProviders();
 
-        protected abstract string GetLanguageName();
+        protected virtual CompletionRules CreateCompletionRules(MostRecentlyUsedList mostRecentlyUsedList)
+        {
+            return new CompletionRules(mostRecentlyUsedList);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Completion/AbstractCompletionService.cs]: Interesting.   I see what you're doing here, and it feels ok.  One thing I think we should do is look at the three completion rules we have (TS/C#/VB), and elevate any common behavior they have (like commit characters up to the base type).

File: src/Features/Core/Completion/AbstractCompletionService.cs [hunk 11]
```diff
@@ -325,34 +276,14 @@ public bool IsTriggerCharacter(SourceText text, int characterPosition, IEnumerab
...
-            {
-                return Task.FromResult(string.Format(FeaturesResources.NoteTabTwiceToInsertTheSnippet, insertionText));
-            }
-
-            return SpecializedTasks.Default<string>();
-        }
+        public abstract Task<TextSpan> GetDefaultTrackingSpanAsync(Document document, int position, CancellationToken cancellationToken);
 
-        public virtual bool SupportSnippetCompletionListOnTab
+        public virtual bool DismissIfEmpty
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Completion/AbstractCompletionService.cs]: => false;

File: src/Features/Core/Completion/ICompletionService.cs [hunk 12]
```diff
@@ -12,6 +12,11 @@ namespace Microsoft.CodeAnalysis.Completion
     internal interface ICompletionService : ILanguageService
     {
         /// <summary>
+        /// The language that this <see cref="ICompletionService"/> targets.
+        /// </summary>
+        string LanguageName { get; }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Completion/ICompletionService.cs]: Why is this necessary in the interface?  Does anyone externally (i.e. not the impl of ICompletionService) use this?

File: src/Features/VisualBasic/Completion/VisualBasicCompletionRules.vb [hunk 13]
```diff
@@ -13,10 +14,22 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Completion
...
+            Public Sub New(mostRecentlyUsedList As MostRecentlyUsedList)
+                MyBase.New(mostRecentlyUsedList)
             End Sub
 
+            ''' <summary>
+            ''' In Turkish Locale, both capital 'i's should be considered same. This behavior matches the compiler behavior.
+            ''' If lowered, both 'i's are lowered to small 'i' with dot
+            ''' </summary>
+            Protected Overrides Function GetCultureSpecificQuirks(candidate As String) As String
+                If CultureInfo.CurrentCulture.Name = "tr-TR" Then
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Completion/VisualBasicCompletionRules.vb]: wowza

File: src/Features/Core/Completion/ICompletionService.cs (hunk 12 above)
Comment (DustinCampbell) [inline src/Features/Core/Completion/ICompletionService.cs]: Yeah, I'm still trying to get rid of it. It gets used way down the line by snippets to determine what language to grab an ```ISnippetInfoService``` for. I still need to dig into that. If I can figure out the right way to get the language name in that case, I'll kill this property.

File: src/Features/VisualBasic/Completion/VisualBasicCompletionRules.vb (hunk 13 above)
Comment (DustinCampbell) [inline src/Features/VisualBasic/Completion/VisualBasicCompletionRules.vb]: Cut and paste :smile:

File: src/Features/Core/Completion/AbstractCompletionService.cs (hunk 9 above)
Comment (DustinCampbell) [inline src/Features/Core/Completion/AbstractCompletionService.cs]: No they're not. I'll be cleaning this up later. Thanks for spotting it!

File: src/Features/Core/Completion/AbstractCompletionService.cs (hunk 10 above)
Comment (DustinCampbell) [inline src/Features/Core/Completion/AbstractCompletionService.cs]: Yup -- I've got a bug to unify the commit characters. I also noticed that we create a new ```CompletionRules``` every time GetCompletionRules() is called, which doesn't seem right since there's very little state in the rules. The only thing that isn't stateless in the rules is a cache of pattern matchers, which I may move elsewhere. As an aside, it's looking like we'll need to make ```PatternMatcher``` public with the completion API. Otherwise, implementing CompletionRules is pretty tough if you want custom behavior.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/DescriptionModifyingCompletionItem.cs (hunk 8 above)
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/DescriptionModifyingCompletionItem.cs]: I'm not 100% sure, but I'll look into it. This is a cut and paste from ```AbstractCompletionService```.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TabKey.cs (hunk 7 above)
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TabKey.cs]: Will do.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Backspace.cs (hunk 6 above)
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Backspace.cs]: I tried separating them at first and it seemed weird because every place this trigger is checked needs to check both. Now that I'm looking at it again, maybe I should go back to separating them?

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/AbstractCSharpCompletionProviderTests.cs (hunk 5 above)
Comment (DustinCampbell) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/AbstractCSharpCompletionProviderTests.cs]: Oops -- thanks!

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Backspace.cs (hunk 6 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Backspace.cs]: I would.  Frankly, I am somewhat surprised that we treat them the same.  (And, if we treat them the same, then why do we have to track if it was an actual delete vs a backspace?).  If they are treated differently somewhere, then we should def have different types IMO.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs (hunk 4 above)
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs]: I'm trying to rationalize the MRU stuff because it's currently baked in a bit wonkily. As an abstraction, it's pretty leaky right now. Still working on it if you have suggestions.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs (hunk 1 above)
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs]: Yeah, it felt weird having these in the main namespace because they're really seldom used. Extenders and consumers almost never need them. However, if you're building something like the Controller, you will.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Backspace.cs (hunk 6 above)
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Backspace.cs]: On the consumption side, we treat them the same in VB. You're right -- that looks a bit weird.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/DescriptionModifyingCompletionItem.cs (hunk 8 above)
Comment (rchande) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/DescriptionModifyingCompletionItem.cs]: When we decide whether to add the "tab twice to insert the snippet" text, we make that decision based on the insertion text. This is basically a hack that makes the CompletionRules pretend the item was committed on tab or double click, to avoid any special casing they might do for committing on open paren or similar.

File: src/Features/Core/Completion/AbstractCompletionService.cs [hunk 14]
```diff
@@ -8,94 +8,57 @@
...
-                // -1 => 1  == Not Found
-                // 0  => 0  == least recently used 
-                // 9  => -9 == most recently used 
-                var index = _committedItems.IndexOf(item.DisplayText);
-                return -index;
-            }
+            return this.MostRecentlyUsedList.GetMRUIndex(item);
         }
 
         public void ClearMRUCache()
```
Comment (jasonmalinowski) [inline src/Features/Core/Completion/AbstractCompletionService.cs]: Same as Cyrus' previous comment: this can be cleaned up too.

File: src/Features/Core/Completion/Snippets/ISnippetCompletionListProvider.cs [hunk 15]
```diff
@@ -0,0 +1,8 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.CodeAnalysis.Completion.Snippets
+{
+    internal interface ISnippetCompletionListProvider
```
Comment (jasonmalinowski) [inline src/Features/Core/Completion/Snippets/ISnippetCompletionListProvider.cs]: Should this inherit from some interface just so the type system can help us out here? It's unclear to me what (if any) types this can be applied to or what such an application means. XML doc comments might just help.

Comment (DustinCampbell) [inline src/Features/Core/Completion/Snippets/ISnippetCompletionListProvider.cs]: This is just a temporary measure. It's a marker interface so we can hide some of the ugliness around snippets right now. Hopefully I'll be able to just remove this later. I'll add a comment to that effect.

File: src/Features/Core/Completion/AbstractCompletionService.cs (hunk 14 above)
Comment (DustinCampbell) [inline src/Features/Core/Completion/AbstractCompletionService.cs]: Yup

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/DescriptionModifyingCompletionItem.cs (hunk 8 above)
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/DescriptionModifyingCompletionItem.cs]: Yeah, that's pretty yucky. :smile:

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs [hunk 16]
```diff
@@ -66,8 +68,20 @@ internal partial class Controller :
             _undoHistoryRegistry = undoHistoryRegistry;
             _allCompletionProviders = allCompletionProviders;
             _autoBraceCompletionChars = autoBraceCompletionChars;
-            _isDebugger = isDebugger;
-            _isImmediateWindow = isImmediateWindow;
+
+            var tagsBuilder = ImmutableArray.CreateBuilder<string>();
+
+            if (isDebugger)
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs]: If you Eventually™ expect this type becomes public, we'll probably need the controller to just have a IEnumerable<string> that comes in. It seems very odd that the two booleans are called out, especially since I think there's only caller of this.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs [hunk 17]
```diff
@@ -162,12 +177,16 @@ private bool StartNewModelComputation(ICompletionService completionService, Comp
...
                 ? GetSnippetCompletionProviders()
                 : GetCompletionProviders();
 
-            sessionOpt.ComputeModel(completionService, triggerInfo, completionProviders, _isDebugger);
-            var filterReason = triggerInfo.TriggerReason == CompletionTriggerReason.BackspaceOrDeleteCommand ? CompletionFilterReason.BackspaceOrDelete : CompletionFilterReason.TypeChar;
+            sessionOpt.ComputeModel(completionService, trigger, completionProviders);
+
+            var filterReason = trigger is BackspaceOrDeleteCharCompletionTrigger
+                ? CompletionFilterReason.BackspaceOrDelete
+                : CompletionFilterReason.TypeChar;
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs]: Doesn't this function also get called during explicit triggers of completion? Is TypeChar the right trigger in this case? Or is this grand unification of "typing forward is the same as explicit trigger" thing I've hoped for?

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs (hunk 16 above)
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs]: I don't expect the Controller to become public. At least not for a long while.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs (hunk 17 above)
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs]: I have no idea what you mean. This was just a simple transformation of the code to longer use the CompletionTriggerReason enum.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs (hunk 2 above)
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs]: We could. In the case of some of the triggers, there's data added as well (e.g. DeletedCharacter).

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Backspace.cs (hunk 3 above)
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Backspace.cs]: Yeah, this is pretty ugly. I'll take a look.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_ReturnKey.cs [hunk 18]
```diff
@@ -64,7 +65,8 @@ private void CommitOnEnter(out bool sendThrough, out bool committed)
             // to the editor.  In single-line debugger windows (Watch, etc), however, we don't
             // want to send the enter though, because those windows don't support displaying
             // more than one line of text.
-            sendThrough = !model.TriggerInfo.IsDebugger || model.TriggerInfo.IsImmediateWindow;
+            sendThrough = !model.Trigger.CustomTags.Contains(WellKnownCompletionTriggerTags.Debugger)
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_ReturnKey.cs]: Are there any custom triggers being used to represent other types of triggering, or is it really being used as "here's a way to pass ambient state along for the ride"? I recognize this might churn stuff, but it very much feels like this CustomTags exists on trigger just because it was the convenient type that happens to be there.

Three interesting, potentially controversial, and likely terrible ideas came to mind while writing that previous sentence:

1. I think you already had some sort of "context" object that was beginning to congeal. Perhaps a type that is the "context" that contains the trigger, various "environment" flags (notably this debugger flag and the interactive flag), and maybe some other pieces of information could become the replacement.
2. Is the debugger flag only being used to control whether we list completion items before their declaration point? And do completion providers get passed an OptionSet from the originating workspace (or can get at it)? If both are "yes", then _perhaps_ we should instead have there be a not-UI-exposed completion option that simply specifies that debugger behavior, and the debugger workspace we create just sets that option explicitly. I could _imagine_ people might actually like that behavior during real typing as well. It's not rare that you might write some code and realize you need to use a local declared farther down. Rather than me breaking context, moving the local up, and then writing the code again, that it's perhaps OK to actually complete it anyways. It's I am _not_ saying we should expose it in a UI (at least not until customers come with pitchforks asking for it :smile:), but if they did come along and ask for that we'd implement that with an option. Given that *also* might fix this problem too, I can't help but notice the coincidence.
3. Rather than having some additional way to shuttle info, this could just be checking the workspace kind. It's still comparing magic strings, but at least without the baggage of extra types.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs (hunk 17 above)
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs]: Oh I see, I didn't see the previous reference to TypeChar since it got cut off due to line length.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_ReturnKey.cs (hunk 18 above)
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_ReturnKey.cs]: @jasonmalinowski 

Those aren't terrible ideas at all! :smile:

>I think you already had some sort of "context" object that was beginning to congeal. Perhaps a type that is the "context" that contains the trigger, various "environment" flags (notably this debugger flag and the interactive flag), and maybe some other pieces of information could become the replacement.

This would need to be a *yet another* type than the "context" object that is passed by the completion service to ```CompletionListProvider.ProduceCompletionListAsync```. I'd prefer not to introduce more of these. However, your point is taken. I was modeling CustomTags off of what we did with diagnostics, but perhaps that's not the best approach here.

>Is the debugger flag only being used to control whether we list completion items before their declaration point? And do completion providers get passed an OptionSet from the originating workspace (or can get at it)? If both are "yes", then perhaps we should instead have there be a not-UI-exposed completion option that simply specifies that debugger behavior, and the debugger workspace we create just sets that option explicitly. I could imagine people might actually like that behavior during real typing as well. It's not rare that you might write some code and realize you need to use a local declared farther down. Rather than me breaking context, moving the local up, and then writing the code again, that it's perhaps OK to actually complete it anyways. It's I am not saying we should expose it in a UI (at least not until customers come with pitchforks asking for it :smile:), but if they did come along and ask for that we'd implement that with an option. Given that also might fix this problem too, I can't help but notice the coincidence.

This is the most appealing idea. Rather than having flags like IsDebugger and IsImmediateWindow that control random behaviors, capturing those behaviors in options is a good idea. I'll try this one out.

>Rather than having some additional way to shuttle info, this could just be checking the workspace kind. It's still comparing magic strings, but at least without the baggage of extra types.

Workspace kind is a pretty big hammer that I'd like to avoid. Of course, you probably listed it last because you'd like me to avoid it too. :smile:

**Comment by DustinCampbell** [comment]
Thanks for all the feedback! That's definitely what I was looking for here. I'll be closing this PR and moving to a new one.
